README.md
# intro #
A connector to use a LEGO Mindstorm ev3 running the ev3dev firmware
(http://www.ev3dev.org) from the Open Roberta lab (http://lab.open-roberta.org).
This is now included by default with ev3dev images (Thanks @dlech), but it is
prevented from running by default, so you do have to enable it once:

    sudo systemctl unmask openrobertalab.service
    sudo systemctl start openrobertalab.service

After running the commands above, it will start automatically after a reboot.
You can turn it back off by running:

    sudo systemctl stop openrobertalab.service
    sudo systemctl mask openrobertalab.service

If the ``openrobertalab`` package is installed and the service is running, the
``Open Roberta Lab`` menu item in brickman will allow you to connect to an Open
Roberta server. This is how the menu will look like:

![Main Menu](/docs/MenuMain.png?raw=true "Main Menu").

Once you selected the ``Open Roberta Lab`` menu item you'll get to this screen:

![Open Roberta Lab](/docs/RobertaLabDisconnected.png?raw=true "Open Roberta Lab").

This offers to connect to the public server as the first item, or to a custom
server as a 2nd item. The 2nd choice is mostly for developers or for using a
local server. When clicking connect, the screen will show a pairing code:

![Pairing Code](/docs/RobertaLabConnecting.png?raw=true "Pairing Code").

This code will have to be entered on the web-ui to establish the link. Once that
has been done a beep-sequence on the ev3 confirms the link and this screen is
shown:

![Connected](/docs/RobertaLabConnected.png?raw=true "Connected").

When a program contains an infinite loop, it can be ``killed`` by pressing
the ``enter`` and ``down`` buttons on the ev3 simultaneously. If this is not
enough to terminate the program, holding the ``back`` button for one second
will kill it, but together with it the connector. The connector will restart
automatically, but one needs to reconnect to the Open Roberta server again.

# build status #

[![Build Status](https://travis-ci.org/OpenRoberta/robertalab-ev3dev.svg?branch=develop)](https://travis-ci.org/OpenRoberta/robertalab-ev3dev/builds)
[![Test Coverage](https://codecov.io/gh/OpenRoberta/robertalab-ev3dev/branch/develop/graph/badge.svg)](https://codecov.io/gh/OpenRoberta/robertalab-ev3dev)


# development #
## prerequisites ##
python3-ev3dev
python3-bluez
python3-dbus
python3-gi

## dist ##

    VERSION="1.3.2" python setup.py sdist

Now you can also build a debian package using ``debuild`` or
``debuild -us -us``. The new package will be in the parent folder.

## upload to ev3 ##
The easiest is to upload the debian package and install it.

    scp ../openrobertalab_1.3.2-1_all.deb maker@ev3dev.local:
    ssh -t maker@ev3dev.local "sudo dpkg --install openrobertalab_1.3.2-1_all.deb;"

Alternatively after changing single files you can do:

    scp roberta/ev3.py robot@ev3dev.local:
    ssh -t robot@ev3dev.local "sudo mv ev3.py /usr/lib/python3/dist-packages/roberta/; sudo systemctl restart openrobertalab"

## configuration ##
The brickman ui will store configuration data under /etc/openroberta.conf. All
configuration can be edited from the UI. If there is a need to manually change
the config, it is advised to stop brickman.

## Testing ##
``python3 -m unittest discover roberta`` or ``nosetests``.
The test require ``python3-httpretty``, but run without ``python3-ev3dev``.

## Logging ##
The service writes status to the system journal.

    sudo journalctl -f -u openrobertalab


debian/changelog
openrobertalab (1.7.4+1.0.0) oldstable; urgency=medium

  * avoid div/0 for the curve block
  * add more parameters to the say block
  * expand test coverage

 -- Stefan Sauer <ensonic@google.com>  Fri, 19 Jan 2018 10:01:07 +0100

openrobertalab (1.7.3+1.0.0) oldstable; urgency=medium

  * support the nxt sound sensor
  * fix joining text with numbers
  * make bluetooth communication interruptible
  * add setLanguage for upcomin say blocks
  * fix timers
  * use a blink thread instead of the broken timer feature

 -- Stefan Sauer <ensonic@google.com>  Fri, 01 Dec 2017 11:53:10 +0100

openrobertalab (1.7.2+1.0.0) stable; urgency=medium

  * fix state handling when canceling the pin-code
  * improve hard/sort resets
  * only set gyro mode when the value changed (to avoid resets)
  * make sound commands interruptible
  * stop motors when stalled

 -- Stefan Sauer <ensonic@google.com>  Mon, 29 May 2017 11:02:24 +0200

openrobertalab (1.7.1+1.0.0) stable; urgency=medium

  * log the python path
  * fix more motor attribute names
  * avoid storing custom properties on monot instances

 -- Stefan Sauer <ensonic@google.com>  Thu, 02 Feb 2017 16:15:28 +0100

openrobertalab (1.7.0+1.0.0) stable; urgency=medium

  * handle more network errors
  * support new image api (allows to pass images as part of the code)
  * use new python sound api

 -- Stefan Sauer <ensonic@google.com>  Wed, 25 Jan 2017 13:53:04 +0100

openrobertalab (1.6.0+1.0.0) stable; urgency=medium

  * rewrite the she-bang since we now use python3
  * add screenshots for readme/wiki
  * fix speed scaling of motors
  * remove use of deprecated speed regullation attribute

 -- Stefan Sauer <ensonic@google.com>  Fri, 25 Nov 2016 20:36:46 +0100

openrobertalab (1.5.0+1.0.0) stable; urgency=medium

  * port to python3

 -- Stefan Sauer <ensonic@google.com>  Mon, 07 Nov 2016 09:29:33 +0100

openrobertalab (1.4.1+1.0.0) stable; urgency=medium

  * add soft abort feature (press and hold enter+down keys)
  * more list operation supported
  * new drive-in-curve block supported in hal
  * two motor drive command are better synced
  * proper gfx<->txt mode switching
  * fix tgb-color sensor values

 -- Stefan Sauer <ensonic@google.com>  Mon, 26 Sep 2016 11:00:54 +0200

openrobertalab (1.4.0+1.0.0) stable; urgency=medium

  * code generator fixes
  * https support
  * more Blocklymethods implemented

 -- Stefan Sauer <ensonic@google.com>  Mon, 25 Apr 2016 16:04:58 +0200

openrobertalab (1.3.2+3.0.0) stable; urgency=medium

  * code generator fixes
  * led blinks are async again
  * better exception reporting
  * handle exceptions when instatiating sensors that are not plugged
  * handle speed update for running motors
  * properly scale sensor values

 -- Stefan Sauer <ensonic@google.com>  Fri, 11 Mar 2016 16:08:10 +0100

openrobertalab (1.3.2+2.0.0) stable; urgency=medium

  * send 'disconnected' when receiving an error
  * drop the 'random' module
  * code style cleanups
  * refactor session code
  * Avoid '01' and 'IO' in the tokens
  * add factory methods for sensor creation
  * Use more portable form of imports.
  * Use the powersupply class from ev3dev-python
  * leds replace flash() api with a manual loop

 -- Stefan Sauer (ensonic) <ensonic@google.com>  Tue, 16 Feb 2016 11:34:54 +0100

openrobertalab (1.3.2+1.0.0) stable; urgency=medium

  [David Lechner]
  * Fix programs not running when started from NEPO.
  * Don't flood the logs
  * Use native package format
  * Drop .py extension from openrobertalab

  [Stefan Sauer]
  * Add automatic tests.
  * Store user programs in home and make them executable.
  * Support release and develop server paths.
  * Switch to the new python-ev3dev api.

 -- David Lechner <david@lechnology.com>  Tue, 29 Dec 2015 22:55:26 -0600

openrobertalab (1.3.2-1) stable; urgency=low

  * update to version 1.3.2

 -- Stefan Sauer (ensonic) <ensonic@google.com>  Tue, 20 Oct 2015 10:11:18 +0200

openrobertalab (1.3.0-1) stable; urgency=low

  * initial package

 -- Stefan Sauer (ensonic) <ensonic@google.com>  Mon, 21 Sep 2015 14:25:33 +0200


roberta/test.py
# Hal and Ev3dev class to satisfy testing

from PIL import Image, ImageDraw


class Hal(object):

    def __init__(self, brickConfiguration, usedSensors=None):
        self.cfg = brickConfiguration

    def clearDisplay(self):
        pass

    def playFile(self, systemSound):
        pass


class Ev3dev(object):
    OUTPUT_A = 'outA'
    OUTPUT_B = 'outB'
    OUTPUT_C = 'outC'
    OUTPUT_D = 'outD'

    INPUT_1 = 'in1'
    INPUT_2 = 'in2'
    INPUT_3 = 'in3'
    INPUT_4 = 'in4'

    class Leds(object):
        BLACK = 0
        GREEN = 1
        RED = 2
        ORANGE = 3
        LEFT = 4
        RIGHT = 5

    Sound = None

    def Button():
        return None

    class PowerSupply(object):
        measured_volts = 0.0

    class Screen(object):
        draw = None

        def __init__(self):
            im = Image.new('1', (178, 128), (0))
            self.draw = ImageDraw.Draw(im)

    class LargeMotor(object):

        def __init__(self, port):
            self.port = port
            self.position = 0
            self.state = False
            self.max_speed = 100
            self.count_per_rot = 360

        def run_to_rel_pos(self, **kwargs):
            self.__dict__.update(kwargs)

        def run_direct(self, **kwargs):
            self.__dict__.update(kwargs)
            # TODO: the calling code waits for the positions to be reached, we
            # only know the directions :/
            if self.duty_cycle_sp >= 0:
                self.position = 10000
            else:
                self.position = -10000

        def run_forever(self, **kwargs):
            self.__dict__.update(kwargs)

        def stop(self):
            pass


roberta/test_ev3.py
import unittest

from .ev3 import Hal
from .test import Ev3dev as ev3dev


class TestHal(unittest.TestCase):
    def test__init__no_cfg(self):
        hal = Hal(None)
        self.assertNotEqual(0, hal.font_w)
        self.assertNotEqual(0, hal.font_h)

    def test__init__simple_cfg(self):
        brickConfiguration = {
            'wheel-diameter': 5.6,
            'track-width': 18.0,
            'actors': {
            },
            'sensors': {
            },
        }
        hal = Hal(brickConfiguration)
        self.assertIn('wheel-diameter', hal.cfg)

    def test__init__with_actor(self):
        brickConfiguration = {
            'wheel-diameter': 5.6,
            'track-width': 18.0,
            'actors': {
                'B': Hal.makeLargeMotor(ev3dev.OUTPUT_B, 'on', 'forward'),
            },
            'sensors': {
            },
        }
        hal = Hal(brickConfiguration)
        self.assertIsNotNone(hal.cfg['actors']['B'])

    def _getStdHal(self):
        brickConfiguration = {
            'wheel-diameter': 5.6,
            'track-width': 18.0,
            'actors': {
                'B': Hal.makeLargeMotor(ev3dev.OUTPUT_B, 'on', 'forward'),
                'C': Hal.makeLargeMotor(ev3dev.OUTPUT_C, 'on', 'forward'),
            },
            'sensors': {
            },
        }
        return Hal(brickConfiguration)

    # rotateRegulatedMotor
    def test_rotateRegulatedMotor_Degree(self):
        hal = self._getStdHal()
        hal.rotateRegulatedMotor('B', 100, 'degree', 90.0)
        self.assertEqual(hal.cfg['actors']['B'].speed_sp, 100)

    def test_rotateRegulatedMotor_Rotations(self):
        hal = self._getStdHal()
        hal.rotateRegulatedMotor('B', 100, 'rotations', 2)
        self.assertEqual(hal.cfg['actors']['B'].position_sp, 720)

    def test_rotateRegulatedMotor_SpeedIsClipped(self):
        hal = self._getStdHal()
        hal.rotateRegulatedMotor('B', 500, 'degree', 90.0)
        self.assertEqual(hal.cfg['actors']['B'].speed_sp, 100)

    # rotateUnregulatedMotor
    def test_rotateUnregulatedMotor_Forward(self):
        hal = self._getStdHal()
        hal.rotateUnregulatedMotor('B', 100, 'power', 100)
        self.assertGreater(hal.cfg['actors']['B'].position, 0)

    def test_rotateUnregulatedMotor_Backward(self):
        hal = self._getStdHal()
        hal.rotateUnregulatedMotor('B', -100, 'power', 100)
        self.assertLess(hal.cfg['actors']['B'].position, 0)

    def test_rotateUnregulatedMotor_SpeedIsClipped(self):
        hal = self._getStdHal()
        hal.rotateUnregulatedMotor('B', 500, 'power', 100)
        self.assertEqual(hal.cfg['actors']['B'].duty_cycle_sp, 100)

    # turnOnRegulatedMotor
    def test_turnOnRegulatedMotor_SpeedIsClipped(self):
        hal = self._getStdHal()
        hal.turnOnRegulatedMotor('B', 500)
        self.assertEqual(hal.cfg['actors']['B'].speed_sp, 100)

    # turnOnUnregulatedMotor
    def test_turnOnUnregulatedMotor_SpeedIsClipped(self):
        hal = self._getStdHal()
        hal.turnOnUnregulatedMotor('B', 500)
        self.assertEqual(hal.cfg['actors']['B'].duty_cycle_sp, 100)

    # stopMotor
    def test_stopMotor_defaultMode(self):
        hal = self._getStdHal()
        hal.stopMotor('B')
        self.assertEqual(hal.cfg['actors']['B'].stop_action, 'coast')

    # stopMotors
    def test_stopMotors(self):
        hal = self._getStdHal()
        hal.stopMotors('B', 'C')
        self.assertEqual(hal.cfg['actors']['B'].stop_action, 'coast')
        self.assertEqual(hal.cfg['actors']['C'].stop_action, 'coast')

    # regulatedDrive
    def test_regulatedDrive(self):
        hal = self._getStdHal()
        hal.regulatedDrive('B', 'C', False, 'forward', 100)
        actors = hal.cfg['actors']
        self.assertEqual(actors['B'].speed_sp, actors['C'].speed_sp)

    # driveDistance
    def test_driveDistance(self):
        hal = self._getStdHal()
        hal.driveDistance('B', 'C', False, 'forward', 100, 10)
        actors = hal.cfg['actors']
        self.assertEqual(actors['B'].speed_sp, actors['C'].speed_sp)
        self.assertEqual(actors['B'].position_sp, actors['C'].position_sp)

    # rotateDirectionRegulated
    def test_rotateDirectionRegulated(self):
        hal = self._getStdHal()
        hal.rotateDirectionRegulated('B', 'C', False, 'right', 100)
        actors = hal.cfg['actors']
        self.assertEqual(actors['B'].speed_sp, -actors['C'].speed_sp)

    # rotateDirectionAngle
    def test_rotateDirectionAngle(self):
        hal = self._getStdHal()
        hal.rotateDirectionAngle('B', 'C', False, 'right', 100, 90.0)
        actors = hal.cfg['actors']
        self.assertEqual(actors['B'].speed_sp, actors['C'].speed_sp)
        self.assertEqual(actors['B'].position_sp, -actors['C'].position_sp)

    # driveInCurve
    def test_driveInCurve_noDist(self):
        hal = self._getStdHal()
        hal.driveInCurve('forward', 'B', 10, 'C', 20)
        actors = hal.cfg['actors']
        self.assertEqual(actors['B'].speed_sp * 2, actors['C'].speed_sp)

    def test_driveInCurve_Dist(self):
        hal = self._getStdHal()
        hal.driveInCurve('forward', 'B', 10, 'C', 20, 100)
        actors = hal.cfg['actors']
        self.assertEqual(actors['B'].speed_sp * 2, actors['C'].speed_sp)
        self.assertGreater(actors['B'].position_sp, 0)
        self.assertGreater(actors['C'].position_sp, 0)

    def test_driveInCurve_ZeroSpeed(self):
        hal = self._getStdHal()
        actors = hal.cfg['actors']
        hal.driveInCurve('forward', 'B', 0, 'C', 0, 100)
        self.assertEqual(actors['B'].speed_sp, 0)
        self.assertEqual(actors['C'].speed_sp, 0)

    def test_driveInCurve_OppositeSpeeds(self):
        hal = self._getStdHal()
        actors = hal.cfg['actors']
        hal.driveInCurve('forward', 'B', 10, 'C', -10, 100)
        self.assertEqual(actors['B'].speed_sp, 10)
        self.assertEqual(actors['C'].speed_sp, -10)


roberta/lab.py
import ctypes
import dbus
import dbus.service
from fcntl import ioctl
import json
import logging
import os
import socket
import stat
import struct
import time
import _thread
import threading
import urllib.request
import urllib.error
import urllib.parse
import sys
# ignore failure to make this testable outside of the target platform
try:
    from ev3dev import auto as ev3dev
    # TODO: prefer the module updated from the server
    # sys.path.prepend(os.path.expanduser('~/.local/python/roberta')
    from .ev3 import Hal
except ImportError:
    from .test import Ev3dev as ev3dev
    from .test import Hal
from .__version__ import version

logger = logging.getLogger('roberta.lab')

# configuration

# TRUE: use a new token per reconnect
# FALSE: try keep using the token for as long as possible
#        (needs robertalab > 1.4 or develop branch)
TOKEN_PER_SESSION = True


# helpers
def getHwAddr(ifname):
    # SIOCGIFHWADDR = 0x8927
    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
        info = ioctl(s.fileno(), 0x8927, struct.pack('256s', ifname[:15]))
    return ':'.join(['%02x' % char for char in info[18:24]])


def generateToken():
    # note: we intentionally leave '01' and 'IO' out since they can be confused
    # when entering the code
    chars = '23456789ABCDEFGHJKLMNPQRSTUVWXYZ'
    # note: we don't use the random module since it is large
    b = os.urandom(8)
    return ''.join(chars[b[i] % len(chars)] for i in range(8))


def getBatteryVoltage():
    return "{0:.3f}".format(ev3dev.PowerSupply().measured_volts)


class Service(dbus.service.Object):
    """OpenRobertab-Lab dbus service

    The status state machine is as follows:

    +-> disconnected
    |   |
    |    v
    +- connected
    |    |
    |    v
    +- registered
    |    ^
    |    v
    +- executing

    """

    def __init__(self, path):
        logger.info('python path: %s\n', (':'.join(sys.path)))
        # passing None for path is only for testing
        if path:
            # needs /etc/dbus-1/system.d/openroberta.conf
            bus_name = dbus.service.BusName('org.openroberta.lab', bus=dbus.SystemBus())
            dbus.service.Object.__init__(self, bus_name, path)
            logger.debug('object registered')
            self.status('disconnected')
        self.hal = Hal(None, None)
        self.hal.clearDisplay()
        self.thread = None
        self.params = {
            'macaddr': '00:00:00:00:00:00',
            'firmwarename': 'ev3dev',
            'menuversion': version.split('-')[0],
        }
        self.updateConfiguration()

    def updateConfiguration(self):
        # or /etc/os-release
        with open('/proc/version', 'r') as ver:
            self.params['firmwareversion'] = ver.read()

        for iface in ['wlan', 'usb', 'eth']:
            for ix in range(10):
                try:
                    ifname = bytes(iface + str(ix), 'ascii')
                    self.params['macaddr'] = getHwAddr(ifname)
                    break
                except IOError:
                    pass
        # reusing token is nice for developers, but the server started to reject
        # them
        if not TOKEN_PER_SESSION:
            self.params['token'] = generateToken()

    @dbus.service.method('org.openroberta.lab', in_signature='s', out_signature='s')
    def connect(self, address):
        logger.debug('connect(%s)', address)
        if self.thread:
            logger.debug('disconnect() old thread')
            # make sure we don't change to disconnected when the thread
            # eventually terminates after the http timeout
            self.thread.service = None
            self.thread.running = False
        # start thread, connecting to address
        self.thread = Connector(address, self)
        self.thread.daemon = True
        self.thread.start()
        # TODO: we have to 'wait' until the connection has been established and
        # we got the token
        # - we could defer the "connected" signal and add another method to get
        #   the code
        self.status('connected')
        return self.thread.params['token']

    @dbus.service.method('org.openroberta.lab')
    def disconnect(self):
        logger.debug('disconnect()')
        self.thread.running = False
        self.status('disconnected')
        # end thread, can take up to 15 seconds (the timeout to return)
        # hence we don't join(), when connecting again we create a new thread
        # anyway
        # self.thread.join()
        # self.status('disconnected')
        self.thread = None

    @dbus.service.signal('org.openroberta.lab', signature='s')
    def status(self, status):
        logger.info('status changed: %s', status)


class GfxMode(object):

    def __init__(self):
        self.tty_name = os.ttyname(sys.stdin.fileno())

    def __enter__(self):
        logger.info('running on tty: %s', self.tty_name)
        with open(self.tty_name, 'r') as tty:
            # KDSETMODE = 0x4B3A, GRAPHICS = 0x01
            ioctl(tty, 0x4B3A, 0x01)

    def __exit__(self, type, value, traceback):
        with open(self.tty_name, 'w') as tty:
            # KDSETMODE = 0x4B3A, TEXT = 0x00
            ioctl(tty, 0x4B3A, 0x00)
            # send Ctrl-L to tty to clear
            tty.write('\033c')


class AbortHandler(threading.Thread):
    """ Key press handler to abort running programms.
        Tests for a center+down press to soft-kill the programm or a 1 sec back
        key press and terminate the whole process"""

    def __init__(self, service, runner):
        threading.Thread.__init__(self)
        self.service = service
        self.running = True
        self.runner = runner

    def run(self):
        self.long_press = 0
        hal = self.service.hal
        while self.running:
            if hal.isKeyPressed('back'):
                logger.debug('back: %d', self.long_press)
                # if pressed for one sec, hard exit
                if self.long_press > 10:
                    logger.info('--- hard abort ---')
                    _thread.interrupt_main()  # throws KeyboardInterrupt
                    self.running = False
                    # something is eating the KeyboardInterrupt, this is a bit
                    # brute force, but works
                    os._exit(1)
                else:
                    self.long_press += 1
            elif hal.isKeyPressed('enter') and hal.isKeyPressed('down'):
                logger.debug('--- soft-abort ---')
                self.running = False
                self.ctype_async_raise(SystemExit)
            else:
                self.long_press = 0
            time.sleep(0.1)

    def __enter__(self):
        self.start()

    def __exit__(self, type, value, traceback):
        self.running = False
        if type is not None:  # an exception has occurred
            logger.debug('Reraising exception: %s', str(type))
            return False      # reraise the exception
        else:
            logger.debug('No exception')
            return True

    def ctype_async_raise(self, exception):
        # adapted from https://gist.github.com/liuw/2407154
        found = False
        target_tid = 0
        for tid, tobj in list(threading._active.items()):
            if tobj is self.runner:
                found = True
                target_tid = tid
                break
        if not found:
            raise ValueError("Invalid thread object")

        ret = ctypes.pythonapi.PyThreadState_SetAsyncExc(ctypes.c_long(target_tid), ctypes.py_object(exception))
        # ref: http://docs.python.org/c-api/init.html#PyThreadState_SetAsyncExc
        if ret == 0:
            raise ValueError("Invalid thread ID")
        elif ret > 1:
            # Huh? Why would we notify more than one threads?
            # Because we punch a hole into C level interpreter.
            # So it is better to clean up the mess.
            ctypes.pythonapi.PyThreadState_SetAsyncExc(target_tid, 0)
            raise SystemError("PyThreadState_SetAsyncExc failed")
        logger.debug("Successfully set asynchronized exception for %d", target_tid)


class Connector(threading.Thread):
    """OpenRobertab-Lab network IO thread"""

    def __init__(self, address, service):
        threading.Thread.__init__(self)
        self.address = address
        self.service = service
        self.home = os.path.expanduser("~")
        if service:
            self.params = service.params
        else:
            self.params = {}
        if TOKEN_PER_SESSION:
            self.params['token'] = generateToken()

        self.registered = False
        self.running = True
        logger.debug('thread created')

    def _store_code(self, filename, code):
        # TODO: what can we do if the file can't be overwritten
        # https://github.com/OpenRoberta/robertalab-ev3dev/issues/26
        # - there is no point in catching if we only log it
        # - once we can report error details to the server, we can reconsider
        #   https://github.com/OpenRoberta/robertalab-ev3dev/issues/20
        with open(filename, 'w') as prog:
            # Apply hotfixes needed until server update
            # - the server generated code is python2 still
            code = code.replace('from __future__ import absolute_import\n', '')
            code = code.replace('in xrange(', 'in range(')
            code = code.replace('#!/usr/bin/python\n', '#!/usr/bin/python3\n')
            prog.write(code)
        os.chmod(filename, stat.S_IXUSR | stat.S_IRUSR | stat.S_IWUSR)
        return code

    def _exec_code(self, filename, code, abort_handler):
        result = 0
        # using a new process would be using this, but is slower (4s vs <1s):
        # result = subprocess.call(["python", filename], env={"PYTHONPATH":"$PYTONPATH:."})
        # logger.info('execution result: %d' % result)
        #
        # NOTE: we don't have to keep pinging the server while running
        #   the code - robot is busy until we send push request again
        #   it would be nice though if we could cancel the running program
        try:
            compiled_code = compile(code, filename, 'exec')
            with abort_handler:
                scope = {
                    '__name__': '__main__',
                    'result': 0,
                }
                exec(compiled_code, scope)
                result = scope['result']
            logger.info('execution finished: result = %d', result)
        except KeyboardInterrupt:
            logger.info("reraise hard kill")
            raise
        except SystemExit:
            result = 143
            logger.info("soft kill")
        except:  # noqa: E722
            result = 1
            # TODO: return exception details as a string and put into a
            # 'nepoexitdetails' field, so that we can show this in the UI
            logger.exception("Ooops:")
        return result

    def _request(self, cmd, headers, timeout):
        url = '%s/%s' % (self.address, cmd)
        while True:
            try:
                logger.debug('sending request to: %s', url)
                req = urllib.request.Request(url, headers=headers)
                data = json.dumps(self.params).encode('utf8')
                logger.debug('  with params: %s', data)
                return urllib.request.urlopen(req, data, timeout=timeout)
            except urllib.error.HTTPError as e:
                if e.code == 404 and '/rest/' not in url:
                    logger.warning("HTTPError(%s): %s, retrying with '/rest'", e.code, e.reason)
                    # upstream changed the server path
                    url = '%s/rest/%s' % (self.address, cmd)
                elif e.code == 405 and not url.startswith('https://'):
                    logger.warning("HTTPError(%s): %s, retrying with 'https://'", e.code, e.reason)
                    self.address = "https" + self.address[4:]
                    url = "https" + url[4:]
                else:
                    raise e
        return None

    def run(self):
        logger.debug('network thread started')
        # network related locals
        # TODO: change the user agent:
        # https://docs.python.org/2/library/urllib2.html#urllib2.Request
        # default is "Python-urllib/2.7"
        headers = {
            'Content-Type': 'application/json'
        }
        timeout = 15  # seconds

        logger.debug('target: %s', self.address)
        while self.running:
            if self.registered:
                self.params['cmd'] = 'push'
                timeout = 15
            else:
                self.params['cmd'] = 'register'
                timeout = 330
            self.params['brickname'] = socket.gethostname()
            self.params['battery'] = getBatteryVoltage()

            try:
                # TODO: what about /api/v1/pushcmd
                # TODO: according to https://tools.ietf.org/html/rfc6202
                # we should use keep alive
                # http://stackoverflow.com/questions/1037406/python-urllib2-with-keep-alive
                # http://stackoverflow.com/questions/13881196/remove-http-connection-header-python-urllib2
                # https://github.com/jcgregorio/httplib2
                response = self._request("pushcmd", headers, timeout)
                reply = json.loads(response.read().decode('utf8'))
                logger.debug('response: %s', json.dumps(reply))
                cmd = reply['cmd']
                if cmd == 'repeat':
                    if not self.registered:
                        self.service.status('registered')
                        self.service.hal.playFile(2)
                    self.registered = True
                    self.params['nepoexitvalue'] = 0
                elif cmd == 'abort':
                    # if service is None, the user canceled
                    if not self.registered and self.service:
                        logger.info('token collision, retrying')
                        self.params['token'] = generateToken()
                        # make sure we don't DOS the server
                        time.sleep(1.0)
                    else:
                        break
                elif cmd == 'download':
                    # TODO: url is not part of reply :/
                    # TODO: we should receive a digest for the download (md5sum) so that
                    #   we can verify the download
                    logger.debug('download code: %s/download', self.address)
                    response = self._request("download", headers, timeout)
                    hdr = response.getheader('Content-Disposition')
                    # save to $HOME/
                    filename = '%s/%s' % (self.home, hdr.split('=')[1] if hdr else 'unknown')
                    code = self._store_code(filename, response.read().decode('utf-8'))
                    logger.info('code downloaded to: %s', filename)
                    # use a long-press of backspace to terminate
                    abort_handler = AbortHandler(self.service, self)
                    abort_handler.daemon = True
                    # This will make brickman switch vt
                    self.service.status('executing')
                    with GfxMode():
                        self.service.hal.clearDisplay()
                        self.params['nepoexitvalue'] = self._exec_code(filename, code, abort_handler)
                        # if the user did wait for a key press, wait for the key for be released
                        # before handing control back (to e.g. brickman)
                        while self.service.hal.isKeyPressed('any'):
                            time.sleep(0.1)
                        self.service.hal.resetState()
                    self.service.status('registered')
                elif cmd == 'update':
                    # FIXME: implement
                    # ensure local module dir
                    # os.mkdirs('os.path.expanduser('~/.local/python/roberta/'))
                    # fetch ev3.py and store to ~/.local/python/roberta/
                    # then restart:
                    # os.execv(__file__, sys.argv)
                    # check if we need to close files (logger?)
                    pass
                else:
                    logger.warning('unhandled command: %s', cmd)
            except urllib.error.HTTPError as e:
                # e.g. [Errno 404]
                logger.error("HTTPError(%s): %s", e.code, e.reason)
                break
            except urllib.error.URLError as e:
                # e.g. [Errno 111] Connection refused
                #                  The handshake operation timed out
                # errors can be nested
                nested_e = None
                if len(e.args) > 0:
                    nested_e = e.args[0]
                elif e.__cause__:
                    nested_e = e.__cause__
                retry = False
                if nested_e:
                    # this happens if packets were lost
                    if isinstance(nested_e, socket.timeout):
                        retry = True
                    # this happens if we loose network
                    if isinstance(nested_e, socket.gaierror):
                        retry = True
                    if isinstance(nested_e, socket.herror):
                        retry = True
                    if isinstance(nested_e, socket.error):
                        retry = True
                else:
                    retry = True

                if not retry:
                    logger.error("URLError: %s: %s", self.address, e.reason)
                    logger.debug("URLError: %s", repr(e))
                    if nested_e:
                        logger.debug("Nested Exception: %s", repr(nested_e))
                    break
            except (socket.timeout, socket.gaierror, socket.herror, socket.error):
                pass
            except:  # noqa: E722
                logger.exception("Ooops:")
        logger.info('network thread stopped')
        if self.service:
            self.service.status('disconnected')
            # don't play if we we just canceled a registration
            if self.registered:
                self.service.hal.playFile(3)
